### PR TITLE
[DO NOT MERGE] rotors_integration

### PR DESCRIPTION
**Pull request by gbmhunter**

### CODE REVIEW

**Comment by LorenzMeier** [comment]
if you feel most things are solid, can you start pushing dependencies like your changes to the optical flow submodule? Please reference any new PRs with a link here so I can find them.

**Comment by jgoppert** [comment]
Just a heads up, I think we should use this instead of rotors gps plugin: https://github.com/PX4/sitl_gazebo/pull/82

**Comment by jgoppert** [comment]
Hopefully  we can sync my gps fix back with rotors.

**Comment by gbmhunter** [comment]
@LorenzMeier , here is the pull request for the `OpticalFlow` repo. Once that is merged in I will update the submodule in the `rotors_integration` branch of `PX4/Firmware` to point back to the correct upstream repo.

**Comment by LorenzMeier** [comment]
Ok, merged, please go ahead and update the submodule so we can get CI to start passing.

**Comment by gbmhunter** [comment]
@LorenzMeier thanks and done.

**Comment by gbmhunter** [comment]
@jgoppert @LorenzMeier are these GPS changes important to get implemented before this pull request is merged with `master`? If not, it might be worth creating a new issue at https://github.com/ethz-asl/rotors_simulator, where we can discuss how your changes can be integrated with the existing GPS plugin.

**Comment by gbmhunter** [comment]
Will look at that `run tests` error in Jenkins tomorrow morning.

**Comment by jgoppert** [comment]
@gbmhunter No need to get the GPS change in now, I will go ahead and post an issue on rotors. It really needs to be cleaned up into a separate module anyways.

**Comment by LorenzMeier** [comment]
Jenkins test this please.

File: Tools/check_submodules.sh [hunk 1]
```diff
@@ -79,4 +79,4 @@ else
 	done
 
 fi
-	exit 0
+	exit 0
```
Comment (LorenzMeier) [inline Tools/check_submodules.sh]: Missing newline

File: Tools/setup_gazebo.bash [hunk 2]
```diff
@@ -19,10 +21,26 @@ BUILD_DIR=$2
...
+fi
+export GAZEBO_MODEL_PATH=${GAZEBO_MODEL_PATH}:${MODEL_DIRECTORY}
+
 # Disabling the remote model download seems only necessary with Gazebo 6
 #export GAZEBO_MODEL_DATABASE_URI=""
-export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:${SRC_DIR}/Tools/sitl_gazebo/Build/msgs/:${BUILD_DIR}/build_gazebo
+
+# Removed the sitl_gazebo msg folder from linker path when integrated rotors_simulation repo instead. May have to add
+# equivalent directory?
+#export LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:${SRC_DIR}/Tools/sitl_gazebo/Build/msgs/:${BUILD_DIR}/build_gazebo
```
Comment (LorenzMeier) [inline Tools/setup_gazebo.bash]: We should not commit commented code.

File: Tools/sitl_run.sh [hunk 3]
```diff
@@ -82,7 +84,10 @@ then
 		# Set the plugin path so Gazebo finds our model and sim
 		source $src_path/Tools/setup_gazebo.bash ${src_path} ${build_path}
 
-		gzserver --verbose ${src_path}/Tools/sitl_gazebo/worlds/${model}.world &
+		# Modified to point to the new worlds folder after sitl_gazebo was replaced with rotors_simulator
+		#gzserver --verbose ${src_path}/Tools/sitl_gazebo/worlds/${model}.world &
```
Comment (LorenzMeier) [inline Tools/sitl_run.sh]: We should not commit commented code

File: integrationtests/python_src/px4_it/mavros/mavros_offboard_attctl_test.py [hunk 4]
```diff
@@ -94,8 +94,10 @@ def test_attctl(self):
         """Test offboard attitude control"""
 
         # FIXME: hack to wait for simulation to be ready
+        print("DEBUG: wait_until_ready() called.\n");
```
Comment (LorenzMeier) [inline integrationtests/python_src/px4_it/mavros/mavros_offboard_attctl_test.py]: Remove debug instrumentation.

File: integrationtests/python_src/px4_it/mavros/mission_test.py [hunk 5]
```diff
@@ -99,6 +99,8 @@ class MavrosMissionTest(unittest.TestCase):
     """
 
     def setUp(self):
+
+        print("DEBUG: MavrosMissionTest.setUp() called.\n");
```
Comment (LorenzMeier) [inline integrationtests/python_src/px4_it/mavros/mission_test.py]: Leftover debug statements.

File: src/firmware/posix/CMakeLists.txt [hunk 6]
```diff
@@ -138,4 +138,4 @@ install(TARGETS px4 DESTINATION ${CMAKE_INSTALL_BINDIR})
 install(DIRECTORY ${PROJECT_SOURCE_DIR}/ROMFS DESTINATION ${CMAKE_INSTALL_DATADIR}/${PROJECT_NAME})
 install(DIRECTORY ${PROJECT_SOURCE_DIR}/posix-configs DESTINATION ${CMAKE_INSTALL_DATADIR}/${PROJECT_NAME})
 
-# vim: set noet ft=cmake fenc=utf-8 ff=unix :
+# vim: set noet ft=cmake fenc=utf-8 ff=unix :
```
Comment (LorenzMeier) [inline src/firmware/posix/CMakeLists.txt]: Missing newline

**Comment by LorenzMeier** [comment]
Nice, almost there!

File: src/firmware/posix/CMakeLists.txt (hunk 6 above)
Comment (gbmhunter) [inline src/firmware/posix/CMakeLists.txt]: This file doesn't have a newline in `master`, vim settings are on last line also. Shall I keep it the way it was?

**Comment by LorenzMeier** [comment]
Every file should end on a new line, so feel free to fix.

File: Tools/check_submodules.sh (hunk 1 above)
Comment (gbmhunter) [inline Tools/check_submodules.sh]: Done.

File: Tools/setup_gazebo.bash (hunk 2 above)
Comment (gbmhunter) [inline Tools/setup_gazebo.bash]: Done.

File: Tools/sitl_run.sh (hunk 3 above)
Comment (gbmhunter) [inline Tools/sitl_run.sh]: Done.

File: integrationtests/python_src/px4_it/mavros/mavros_offboard_attctl_test.py (hunk 4 above)
Comment (gbmhunter) [inline integrationtests/python_src/px4_it/mavros/mavros_offboard_attctl_test.py]: Done.

File: integrationtests/python_src/px4_it/mavros/mission_test.py (hunk 5 above)
Comment (gbmhunter) [inline integrationtests/python_src/px4_it/mavros/mission_test.py]: Done.

File: src/firmware/posix/CMakeLists.txt (hunk 6 above)
Comment (gbmhunter) [inline src/firmware/posix/CMakeLists.txt]: Done.

**Comment by gbmhunter** [comment]
Oh interesting, rebasing also detaches all the comments in the pull-request, I guess that makes sense, it is rewriting all the commit hashes...

**Comment by jgoppert** [comment]
Why don't we just have a separate official px4_sim ROS package. I think this is getting out of hand with all of the submodules.

File: Tools/setup_gazebo.bash [hunk 7]
```diff
@@ -7,6 +7,8 @@
 #
 # License: according to LICENSE.md in the root directory of the PX4 Firmware repository
 
+echo "setup_gazebo.bash called."
```
Comment (jgoppert) [inline Tools/setup_gazebo.bash]: Cleanup.

File: integrationtests/run_tests.bash [hunk 8]
```diff
@@ -31,16 +33,61 @@ while getopts "h?og" opt; do
...
 # determine the directory of the source given the directory of this script
 pushd `dirname $0` > /dev/null
 SCRIPTPATH=`pwd`
 popd > /dev/null
 ORIG_SRC=$(dirname $SCRIPTPATH)
 
+# =============================================================================================== #
+# =================================== Install/build OpticalFlow ================================= #
+# =============================================================================================== #
+echo "=====> Building/installing OpticalFlow"
```
Comment (jgoppert) [inline integrationtests/run_tests.bash]: I don't like the idea of adding all of these submodules. See general comment on thread.

File: integrationtests/run_tests.bash [hunk 9]
```diff
@@ -79,17 +126,42 @@ else
...
-catkin_make
+
+# These build parameters are used by the CMakeLists.txt in the
+# rotors_gazebo_plugins and rotors_gazebo packages.
+# ADDITIONAL_INCLUDE_DIRS=...				This assumes mav_msgs has been symlinked to the catkin workspace as per above
+# BUILD_MAVLINK_INTERFACE_PLUGIN=TRUE		Build the MAVLink interface plugin for Gazebo
+# BUILD_OCTOMAP_PLUGIN=FALSE				Do not build the Octomap plugin for Gazebo
+# BUILD_OPTICAL_FLOW_PLUGIN=TRUE			Build the optical flow plugin for Gazebo
+# NO_ROS=TRUE 								Instruct the Gazebo plugins to build without ROS dependencies. Note that even though ROS is used for these tests, PX4 uses the Gazebo plugins without any ROS dependencies.
+# CMAKE_MODULE_PATH=$OPTICAL_FLOW_INSTALL_DIR		Tell catkin where to find the OpticalFlow installation, so find(OpticalFlow) works.
```
Comment (jgoppert) [inline integrationtests/run_tests.bash]: One benefit of the existing catkin workspace integration test was to make sure the px4 package still operates normally in the ROS environment for SITL. So building with no-ros isn't helpful. The catkin workspace used to be fully functional in ROS. Now I get a bunch of references to rotors packages that aren't cloned when trying  to build. If you go this route, can you clone all of the required packages and test the ROS functionality.

Comment (jgoppert) [inline integrationtests/run_tests.bash]: @gbmhunter  @devbharat @LorenzMeier I think we need to discuss how to do this cleanly.  What you have done here breaks one of my major use cases.

**Comment by jgoppert** [comment]
I'm trying to upstream the rotors functionality that we need to gazebo natively. Here is my first patch to support a random walk noise model: https://bitbucket.org/osrf/gazebo/pull-requests/2670. So this might take awhile. I'm OK with getting this in, if that is the best approach for now.

**Comment by nicolaerosia** [comment]
@jgoppert any updates?

**Comment by TSC21** [comment]
@gbmhunter what's the status of this?

**Comment by gbmhunter** [comment]
@TSC21 I'm not sure anymore! About half a year ago I was preparing this merge request and believed to have gotten it somewhere close to mergable. My 2-month term then ended and I had to move on...I believe @jgoppert found some usage issues just before I left, I'm not sure on the progress made to fix that.

**Comment by TSC21** [comment]
> @TSC21 I'm not sure anymore! About half a year ago I was preparing this merge request and believed to have gotten it somewhere close to mergable. My 2-month term then ended and I had to move on...I believe 

Are you able to sync this up again so to push this again to master? I think it would be worth for us all to have `sitl_gazebo` sync with rotors.

**Comment by gbmhunter** [comment]
@TSC21 Sorry I am currently living in Canada and have a full-time job, I'm not really in the position to take on more work right now.

**Comment by TSC21** [comment]
@gbmhunter understandable. np

**Comment by TSC21** [comment]
@jgoppert maybe when we have time we consider in getting this forward?

**Comment by dagar** [comment]
Closing as stale.

**Pull request closed without merging**